### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by joshday**
As part of trying to understand the code better, I ended up rewriting it 😅.

### My goals were:

1. Simplification
2. Readability
3. Moving the AMR "closer" to the operations code.  This helps with a lot, including making it easier to support the "observables" part of the AMR.

The whole package is currently a single ~350-line file: SimulationService.jl.  I'd expect this to grow/split into more files as needed in the future.  For now it's nice for searchability and getting new people involved in the code.

---

### TODOs for parity with current version

- status endpoint
- calibrate endpoint
- ensemble endpoint

These should be easy lifts, using the `Simulate` code as a template.  For example, here is all of the simulate-specific code:

```Julia
struct Simulate <: Operation
    sys::ODESystem
    timespan::Tuple{Float64, Float64}
end

Simulate(o::OperationRequest) = Simulate(ode_system_from_amr(only(o.amr)), o.timespan)

function solve(op::Simulate; kw...)
    prob = ODEProblem(op.sys, [], op.timespan, saveat=1)
    sol = solve(prob; progress = true, progress_steps = 1, kw...)
    DataFrame(sol)
end
```

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -5,39 +5,23 @@ version = "0.6.0"
...
 AMQPClient = "79c8b4cd-a41a-55fa-907c-fab5288e1383"
-AlgebraicPetri = "4f99eebe-17bf-4e98-b6a1-2c4f205a959b"
-Bijections = "e2ed5e7c-b2de-5872-ae92-c73ca462fb04"
-CSV = "336ed68f-0bac-5ca0-87d4-7b16caf5d00b"
-Catlab = "134e5e36-593f-5add-ad60-77f754baafbe"
 DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
+Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
 DifferentialEquations = "0c46a032-eb83-5123-abaf-570d42b7fbaa"
-Downloads = "f43a241f-c20a-4ad4-852c-f6b1247861c6"
-EasyModelAnalysis = "ef4b24a4-a090-4686-a932-e7e56a5a83bd"
```
Comment (ChrisRackauckas) [inline Project.toml:15]: how are you doing the calibration without EMA?

File: src/SimulationService.jl [hunk 2]
```diff
@@ -1,77 +1,352 @@
...
+
+        body = JSON3.write((; simulation_id = o.job_id))
+        return Response(201, ["Content-Type" => "application/json; charset=utf-8"]; body=body)
+    catch ex
+        return Response(500, ["Content-Type" => "application/json; charset=utf-8"]; body=JSON3.write((; error=string(ex))))
+    end
+end
+
+function get_operation_type(op::String)
+    d = Dict(lowercase(string(T)) => T for T in subtypes(Operation))
```
Comment (JeffBezanson) [inline src/SimulationService.jl:313]: This should be a constant; `subtypes` is very slow and requires `InteractiveUtils`.

File: src/SimulationService.jl [hunk 3]
```diff
@@ -1,77 +1,389 @@
...
+    statenames = [Symbol(s.id) for s in model.states]
+    statevars  = [only(@variables $s) for s in statenames]
+    statefuncs = [only(@variables $s(t)) for s in statenames]
+
+    # get parameter values and state initial values
+    paramnames = [Symbol(x.id) for x in ode.parameters]
+    paramvars = [only(@parameters $x) for x in paramnames]
+    paramvals = [x.value for x in ode.parameters]
+    sym_defs = paramvars .=> paramvals
+    initial_exprs = [MathML.parse_str(x.expression_mathml) for x in ode.initials]
```
Comment (ChrisRackauckas) [inline src/SimulationService.jl:120]: @jpfairbanks this is where we use the MathML, but we could just symbolics parse the other expressions.

**Comment by joshday** [comment]
Okay, still needs to get the calibrate endpoints back online to achieve parity with previous version, but other than that I think this is good to go.  

I'll add it back in tomorrow morning.

**Comment by joshday** [comment]
I still need to bring calibrate online, but other than that, I think we're good to go.

For reviewers, the things that need the most scrutiny are functions that contain a mention of `SIMSERVICE_ENABLE_TDS`.

- When `SIMSERVICE_ENABLE_TDS = false`, we just print out the equivalent of what would've been sent to the TDS.  So this I can test locally, but not the `SIMSERVICE_ENABLE_TDS=true` case (unless this is possible by setting the `SIMSERVICE_TDS_URL` variable?).

File: src/SimulationService.jl [hunk 4]
```diff
@@ -1,77 +1,472 @@
...
+function solve(op::Simulate; kw...)
+    # joshday: What does providing `u0 = []` do?  Don't we know what u0 is from AMR?
+    prob = ODEProblem(op.sys, [], op.timespan, saveat=1)
+    sol = solve(prob; progress = true, progress_steps = 1, kw...)
+    DataFrame(sol)
+end
+
+#-----------------------------------------------------------------------------# calibrate
+struct Calibrate <: Operation
+    # TODO
```
Comment (ChrisRackauckas) [inline src/SimulationService.jl:460]: the main thing I'd need to know is what exactly would be in here.

File: src/SimulationService.jl [hunk 5]
```diff
@@ -1,77 +1,472 @@
...
+end
+
+
+#-----------------------------------------------------------------------------# simulate
+struct Simulate <: Operation
+    sys::ODESystem
+    timespan::Tuple{Float64, Float64}
+end
+
+Simulate(o::OperationRequest) = Simulate(ode_system_from_amr(o.model), o.timespan)
```
Comment (ChrisRackauckas) [inline src/SimulationService.jl:449]: out of curiosity, where it is known that OperationRequest has o.model and o.timespan?

File: src/SimulationService.jl (hunk 4 above)
Comment (joshday) [reply src/SimulationService.jl:461]: Anything you want.  It just needs to have a `Calibrate(o::OperationRequest)` method.

File: src/SimulationService.jl (hunk 5 above)
Comment (joshday) [reply src/SimulationService.jl:450]: I'm not sure I follow.  Those are fields of the struct.

Comment (joshday) [reply src/SimulationService.jl:450]: Check `examples/` to see what the incoming requests look like.  E.g.

```json
{
  "model_config_id": "f20ea672-bfa4-422b-884f-0a4c09356fd0",
  "timespan": {"start": 0, "end": 90}
}
```

We then reach out to the TDS to get the the AMR (`model`) from the given `model_config_id`.

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**